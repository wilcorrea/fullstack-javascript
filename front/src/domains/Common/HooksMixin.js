/**
 * @mixin HooksMixin
 */
export default {
  /**
   * The hooks!
   */
  hooks: {},
  /**
   */
  methods: {
    /**
     * @param hooks
     */
    installHooks (hooks) {
      this.$options.hooks = hooks
    },
    /**
     * @param {string} hook
     */
    triggerHook (hook) {
      if (typeof this.$options.hooks[hook] === 'function') {
        this.$options.hooks[hook].call(this)
      }
    }
  },
  /**
   */
  created () {
    this.triggerHook('component:created')
  },
  /**
   */
  mounted () {
    this.triggerHook('component:mounted')
  },
  /**
   */
  destroyed () {
    this.triggerHook('component:destroyed')
  }
}
