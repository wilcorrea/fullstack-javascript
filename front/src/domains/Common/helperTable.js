import { get } from 'src/app/Util/general'
import $lang from 'src/lang'
import { breakLine } from 'src/app/Util/string'

/**
 * @param color
 * @returns {Object}
 */
export const style = (color) => ({
  'border-left': `2px solid ${color}`,
  'text-transform': 'uppercase',
  'padding': '0px 0px 0px 5px',
  'font-size': '0.7em',
  'color': 'rgb(60, 73, 93)',
  'line-height': '120%'
})

/**
 * @param {Object} values
 * @param {string} value
 * @param {string} lang
 * @param {boolean} broken
 * @returns {function({h: *, record: *}): *[]}
 */
export const status = (values, value, lang, broken = true) => {
  /**
   * @param h
   * @param record
   * @returns {*}
   */
  return ({ h }) => {
    const answer = (color, content) => {
      const data = {
        domProps: {
          innerHTML: broken ? breakLine(String(content)) : content
        },
        style: style(color)
      }
      return [h('div', data)]
    }
    return answer(get(values, value, {}), $lang(`${lang}.${value}`, value))
  }
}
