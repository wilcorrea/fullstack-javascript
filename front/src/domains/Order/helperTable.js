import { status as statusCommon } from 'src/domains/Common/helperTable'
import { get } from 'src/app/Util/general'
import { money } from 'src/app/Util/currency'
import { color } from 'src/app/Util/ui'

import { statusBackgrounds } from './helper'

let random

/**
 * @param value
 * @returns {string}
 */
export const amount = ({ value }) => money(value)

/**
 * @returns {Object}
 */
export const paymentMethods = () => ({
  CREDIT_CARD: 'credit_card',
  DEBIT_CARD: 'sd_card',
  BOLETO: 'receipt',
  ONLINE_BANK_FINANCING: 'account_balance',
  ONLINE_BANK_DEBIT: 'account_balance_wallet',
  WALLET: 'attach_money'
})

/**
 * @param {CreateElement} h
 * @param record
 * @returns {Array<VNode>}
 */
export const payments = ({ h, record }) => {
  const method = get(record, 'payments.[0].fundingInstrument.method')
  return [h('app-icon', { props: { name: get(paymentMethods(), method, 'warning') } })]
}

/**
 * @param {CreateElement} h
 * @param record
 * @returns {Array<VNode>}
 */
export const status = ({ h, value }) => {
  const lang = 'domains.order.status'
  return statusCommon(statusBackgrounds(), value, lang)({ h })
}

/**
 * @param {CreateElement} h
 * @param {any} value
 * @returns {Array<VNode>}
 */
export const customer = ({ h, value }) => {
  const style = {
    'height': '30px',
    // 'display': 'flex',
    // 'align-items': 'center',
    // 'justify-content': 'center',
    'margin-left': '14px',
    'max-width': '18vw',
    'overflow': 'hidden',
    'text-overflow': 'ellipsis',
    'white-space': 'nowrap',
    'padding': '10px 0 0 0'
  }
  if (random === undefined) {
    // random = confirm('Generate random colors to user avatars?')
  }
  const name = get(value, 'fullname')
  const props = {
    letter: name.substring(0, 1),
    background: random ? color() : undefined,
    color: random ? color() : undefined
  }
  return [
    h('app-avatar', { props }),
    h('div', { style }, name)
  ]
}
