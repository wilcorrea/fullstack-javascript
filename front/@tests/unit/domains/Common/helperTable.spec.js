import { shallowMount } from '@vue/test-utils'
import { status, style } from 'src/domains/Common/helperTable'

describe('helperTable', () => {
  const values = { yes: 'green', no: 'red' }
  const lang = ':/'
  const wrapper = (value) => shallowMount({
    render (h) {
      return status(values, value, lang)({ h })
    }
  })

  /**
   * @test
   */
  it('status composition function: yes', async () => {
    const component = wrapper('yes')
    const html = component.html()
    const styles = style('green')
    Object.keys(styles).forEach((key) => {
      expect(html).toMatch(`${key}: ${styles[key]}`)
    })
    expect(component.text()).toMatch('yes')
    expect(html).toMatchSnapshot('yes')
  })

  /**
   * @test
   */
  it('status composition function: no', async () => {
    const component = wrapper('no')
    const html = component.html()
    const styles = style('red')
    Object.keys(styles).forEach((key) => {
      expect(html).toMatch(`${key}: ${styles[key]}`)
    })
    expect(component.text()).toMatch('no')
    expect(html).toMatchSnapshot('no')
  })
})
