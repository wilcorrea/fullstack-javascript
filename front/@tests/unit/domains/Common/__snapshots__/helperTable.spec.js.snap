// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helperTable status composition function: no: no 1`] = `<div style="border-left: 2px solid red; text-transform: uppercase; padding: 0px 0px 0px 5px; font-size: 0.7em; color: rgb(60, 73, 93); line-height: 120%;">no</div>`;

exports[`helperTable status composition function: yes: yes 1`] = `<div style="border-left: 2px solid green; text-transform: uppercase; padding: 0px 0px 0px 5px; font-size: 0.7em; color: rgb(60, 73, 93); line-height: 120%;">yes</div>`;
